### PR TITLE
Rework apis

### PR DESCRIPTION
**Pull request by dlongley**

### CODE REVIEW

File: AccessManager.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+export default class AccessManager {
+  /**
+   * Creates a new instance of an AccessManager. Should only be called
+   * by ProfileManager.
+   *
+   * @param {Object} options - The options to use.
+   * @param {Object} options.profile - The profile to manage access for.
+   * @param {string} options.profileManager - The parent `profileManager`
+   *  instance.
+   * @param {Object} options.users - A `users` Collection instance.
```
Comment (mattcollier) [inline AccessManager.js:15]: Not sure *yet* what a users collection instance is, but maybe this should be `usersCollection`.

File: AccessManager.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    // store capabilities for accessing the profile agent's user document and
+    // the kak in the profileAgent record in the backend
+    await profileService.updateAgentCapabilitySet({
+      account: accountId,
+      profileAgentId,
+      // this map includes capabilities for user document and kak
+      zcaps: agentRecordZcaps
+    });
+
+    return {...agentDoc.content};
```
Comment (mattcollier) [inline AccessManager.js:111]:
```suggestion
    return agentDoc.content;
```

File: Collection.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
```
Comment (mattcollier) [inline Collection.js:4]:
```suggestion
```

File: Collection.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      throw new TypeError(`"item" must be an object.`);
+    }
+    if(!(item.type === this.type ||
+      (Array.isArray(item.type) && item.type.includes(this.type)))) {
+      throw new TypeError(
+        `"item.type" (${item.type}) must include "${this.type}".`);
+    }
+    const edvDoc = await this._getEdvDocument();
+    const id = await EdvClient.generateId();
+    return await edvDoc.write({doc: {id, content: item, meta}});
```
Comment (mattcollier) [inline Collection.js:29]:
```suggestion
    return edvDoc.write({doc: {id, content: item, meta}});
```

File: Collection.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  async get({id} = {}) {
+    if(typeof id !== 'string') {
+      throw new TypeError('"id" or must be given.');
```
Comment (mattcollier) [inline Collection.js:34]:
```suggestion
      throw new TypeError('"id" must be a document ID.');
```

File: Collection.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    if(results.length > 0) {
+      return results[0];
+    }
+    return null;
+  }
+
+  async getAll() {
+    const {instance, capability} = this;
+    const results = await this._findDocuments(
+      {instance, type: this.type, capability});
```
Comment (mattcollier) [inline Collection.js:46]:
```suggestion
    const {instance, capability, type} = this;
    const results = await this._findDocuments({instance, type, capability});
```

File: Collection.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    const existing = await this.get({id: item.id});
+    const edvDoc = await this._getEdvDocument({id: existing.id});
+    const updatedDoc = {
+      ...existing
+    };
+    updatedDoc.content = item;
+    if(meta) {
+      updatedDoc.meta = meta;
+    }
+    return await edvDoc.write({
```
Comment (mattcollier) [inline Collection.js:68]:
```suggestion
    return edvDoc.write({
```

File: Collection.js [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      doc: updatedDoc
+    });
+  }
+
+  async remove({id} = {}) {
+    if(typeof id !== 'string') {
+      throw new TypeError('"id" or must be given.');
```
Comment (mattcollier) [inline Collection.js:75]:
```suggestion
      throw new TypeError('"id" must be a document ID.');
```

File: ProfileManager.js [hunk 9]
```diff
@@ -422,121 +448,187 @@ export default class ProfileManager {
...
+        capability: userHmac,
+        invocationSigner
+      })
     });
-    return updatedUserDoc.content;
+    for(const index of indexes) {
+      edvClient.ensureIndex(index);
+    }
+    const users = new Collection(
+      {type: 'User', client: edvClient, capability, invocationSigner});
```
Comment (mattcollier) [inline ProfileManager.js:519]: possibly just make `client` be `edvClient`.  And maybe Collection should be EdvCollection?

File: Collection.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
+/*!
+ * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
```
Comment (dlongley) [inline Collection.js:5]:
```suggestion
```

File: Collection.js (hunk 5 above)
Comment (dlongley) [inline Collection.js:34]:
```suggestion
      throw new TypeError('"id" or must be a string.');
```

File: Collection.js (hunk 8 above)
Comment (dlongley) [inline Collection.js:75]:
```suggestion
      throw new TypeError('"id" must be a string.');
```

File: AccessManager.js (hunk 1 above)
Comment (dlongley) [reply AccessManager.js:15]: I'm happy with the shorter name if you're amenable.

File: ProfileManager.js (hunk 9 above)
Comment (dlongley) [reply ProfileManager.js:519]: I'm fine with changing to `client`. Don't want to change `Collection` right now.

File: Collection.js (hunk 8 above)
Comment (dlongley) [reply Collection.js]: Updated; it's also not a doc ID -- it's an ID for the item or content. I went with the general pattern of just saying it needs to be a string.

File: ProfileManager.js (hunk 9 above)
Comment (mattcollier) [reply ProfileManager.js:519]: I'm suggesting that the param to the Collection constructor be `edvClient` etc.  This is just one instance where `edvClient` is being assigned to `client` in this API.  Is that the change you are amenable to?

Comment (dlongley) [reply ProfileManager.js:519]: I'm fine with calling the param `edvClient` inside Collection.

**Pull request merged**